### ISSUE TITLE
create irodsMonitor binary

### ISSUE THREAD

**Issue by trel** [irods/irods#3610]
- should be started/stopped/managed by init.d script, like irodsServer
- only runs alongside catalog provider(s)?  or alongside all servers?  contention?  truth?
- should check heartbeat of all known irodsServers in the Zone
  - `server_config.json` setting of 'monitor_servers_delay_in_seconds'
    - default delay between heartbeat checks set to 1m?
  - upon good heartbeat
    - potentially perform small test put/get?
    - when marked as 'down_by_irodsMonitor', set to 'up'
  - upon bad heartbeat or bad test put/get, when marked as 'up'
    - mark resource as 'down_by_irodsMonitor'
    - consider appending a timestamp... 'down_by_irodsMonitor-YYYYMMDDTHHmmss.ffffffZ'
- write all activity to log?   only some activity?   configuration for this setting?

**Comment by trel** [comment]
if the above scenario is option 1...

option 2)
- an irodsMonitor on `n` catalog providers (consider an HA deployment)
- but this will potentially cause contention between monitors and generate `n` times the traffic to each other server

so option 3)
- raft style distributed consensus algorithm and run irodsMonitor binary on every server
- would require a new port for communication

so option 4)
- instead of irodsMonitor binary and new process everywhere...
- just have a new thread as part of existing irodsServer on every server
- hit a new API endpoint on the configured catalog provider from every server
- no new port necessary, runs on 1247
- server_config.json holds `heartbeat_delay_in_seconds`

**Comment by tsmeele** [comment]
Some items to consider:
- scalable, secure monitoring solutions allow servers to push their data to a monitoring host, rather than data being pulled centrally
- the monitoring data is not just up/down status info, but could also include quantative info such as number of concurrent users/processes, number of jobs on deferred rule queue, storage resource capacity left, cpu load %, etc
- irods servers should be able to provide the measurement data, yet the actual monitoring/analyze/visualization/exception reporting activities could be realized in close cooperation with any of a variety of existing tools (Nagios, Icinga, Zabbix, Prometheus, ...).

**Comment by trel** [comment]
RFC discussion at
https://github.com/irods/irods_rfcs/blob/master/0002_irodsmonitor.md

**Comment by jasoncoposky** [comment]
Consider a [gossip protocol](https://en.wikipedia.org/wiki/Gossip_protocol)

**Comment by RubenGarcia** [comment]
Hello! Is there a draft timeline about this? I see things are moving since some issues are referencing this.

**Comment by trel** [comment]
Not really a timeline - we're just continuing an in-the-public refactoring of the expectations / goals of these different pieces.   The new asynchronous/cron capability of the main server may consume some of the functionality defined in this issue.

Is there something in particular that is useful/helpful to you here that you're most interested in?

**Comment by RubenGarcia** [comment]
I found this page due to the log line (every 10 seconds) which prometheus creates when checking iRODS:
<date> pid:<pid>remote addresses: <addr> ERROR: readWorkerTask - readStartupPack failed. -4000
I wanted to get rid of the log line while still allowing prometheus to query irods state.

My plan is to create an iRODS prometheus exporter with only one metric (server is up or down), to get rid of the line.
Once this irodsMonitor is working, my prometheus exporter can be discarded and the irodsMonitor used instead.

**Comment by trel** [comment]
That line is due to a 'malformed' StartupPack.

Try this, and the error will no longer appear in the serverLog.

https://docs.irods.org/4.2.10/system_overview/tips_and_tricks/#monitoring-status-of-irods-servers

**Comment by RubenGarcia** [comment]
Yes, that was the way I planned to query iRODS within the prometheus exporter. Thanks for the thumbs up.

**Comment by RubenGarcia** [comment]
https://github.com/lexis-project/-irods-prometheus-exporter

**Comment by trel** [comment]
Nice - that's a nice ecosystem.
